checkstyle/src/separate.rs

```diff
@@ -0,0 +1,174 @@
+        Ok(content) => Some(content),
+        Err(err) => {
+            if err.kind() == std::io::ErrorKind::InvalidData {
+                eprintln!(
+                    "Warning: File '{}' contains invalid UTF-8 data, skipping...",
+                    file_path.display()
+                );
+                None
+            } else {
+                eprintln!("Failed to read file '{}': {}", file_path.display(), err);
+                None
+            }
+        }
+    }
+}
+
+pub fn check_matching(file: &str, regex: &str) -> bool {
+    let re = Regex::new(&format!("(?m){}", regex)).unwrap();
+    re.is_match(file)
+}
+
+fn list_files_in_folder(folder_path: &Path) -> Result<Vec<PathBuf>, std::io::Error> {
+    let mut file_paths = Vec::new();
+
+    for entry in fs::read_dir(folder_path)? {
+        let entry = entry?;
+        let path = entry.path();
+
+        if path.is_symlink() {
+            continue;
+        }
+
+        if path.is_file() {
+            file_paths.push(path.clone());
+        } else if path.is_dir() {
+            if let Ok(sub_paths) = list_files_in_folder(&path) {
+                file_paths.extend(sub_paths);
+            }
+        }
+    }
+
+    Ok(file_paths)
+}
+
+fn all_files(folder_path: &Path) -> Vec<PathBuf> {
+    list_files_in_folder(folder_path).unwrap_or_default()
+}
+
+fn files_matching_patterns(
+    target_path: &Path,
+    include_patterns: Option<Vec<String>>,
+    exclude_patterns: Option<Vec<String>>,
+) -> Vec<PathBuf> {
+    let target_file_paths = all_files(target_path);
+    let include_globs = include_patterns.unwrap_or(vec!["**/*".to_string()]);
+    let exclude_globs = exclude_patterns.unwrap_or(Vec::new());
+    target_file_paths
+        .into_iter()
+        .filter(|target_file_name| {
+            let file_name = target_file_name.file_name().unwrap().to_str().unwrap();
+
+            let include_match = include_globs.iter().any(|include_glob| {
+                let pattern = Pattern::new(include_glob).unwrap();
+                pattern.matches(file_name)
+            });
+
+            let exclude_match = exclude_globs.iter().any(|exclude_glob| {
+                let pattern = Pattern::new(exclude_glob).unwrap();
+                pattern.matches(file_name)
+            });
+
+            include_match && !exclude_match
+        })
+        .collect()
+}
+
+fn nonmatching_files_from_list(regex_file_path: &Path, file_paths: Vec<PathBuf>) -> Vec<PathBuf> {
+    let regex = match read_file_content(regex_file_path) {
+        Some(regex_content) => regex_content,
+        None => return Vec::new(),
+    };
+
+    file_paths
+        .into_iter()
+        .filter(|target_file_name| {
+            if let Some(content) = read_file_content(target_file_name.as_ref()) {
+                !check_matching(&content, &regex)
+            } else {
+                false
+            }
+        })
+        .collect()
+}
+
+pub fn separate_regex_matching_files(
```

Comment:
'Separate' as a term makes more sense if we were to return two lists of paths (matching and non-matching). However, I don't think we want this function. Let's move the contents into the main function.